onnxruntime/python/dlpack_convertor.cc

```diff
@@ -0,0 +1,234 @@
+          return DataTypeImpl::GetType<int32_t>();
+        case 64:
+          return DataTypeImpl::GetType<int64_t>();
+        default:
+          ORT_THROW("Unsupported kInt bits " + std::to_string(dtype.bits));
+      }
+    case DLDataTypeCode::kDLFloat:
+      switch (dtype.bits) {
+        case 16:
+          return DataTypeImpl::GetType<MLFloat16>();
+        case 32:
+          return DataTypeImpl::GetType<float>();
+        case 64:
+          return DataTypeImpl::GetType<double>();
+        default:
+          ORT_THROW("Unsupported kFloat bits " + std::to_string(dtype.bits));
+      }
+    default:
+      ORT_THROW("Unsupported code " + std::to_string(dtype.code));
+  }
+}
+
+const char* GetOrtDeviceName(const OrtDevice& device) {
+  switch (device.Type()) {
+    case OrtDevice::CPU:
+      return CPU;
+    case OrtDevice::GPU:
+      return CUDA;
+    default:
+      ORT_THROW("Unknown device type: ", device.Type());
+  }
+}
+
+bool IsContiguousTensor(const DLTensor& tensor) {
+  if (!tensor.strides) {
+    return true;
+  }
+
+  int64_t running_size = 1;
+  for (int i = tensor.ndim - 1; i >= 0; i--) {
+    if (tensor.strides[i] != running_size) {
+      return false;
+    }
+
+    running_size *= tensor.shape[i];
+  }
+
+  return true;
+}
+
+}  // namespace
+
+// This function returns a shared_ptr to memory managed DLpack tensor
+// constructed out of OrtValue.
+DLManagedTensor* OrtValueToDlpack(const OrtValue& ort_value) {
+  ORT_ENFORCE(ort_value.IsTensor(), "Only tensor type OrtValues are supported");
+  OrtDLManagedTensor* ort_dlmanaged_tensor(new OrtDLManagedTensor);
+  const Tensor& tensor = ort_value.Get<Tensor>();
+  ort_dlmanaged_tensor->handle = ort_value;
+  ort_dlmanaged_tensor->tensor.manager_ctx = ort_dlmanaged_tensor;
+  ort_dlmanaged_tensor->tensor.deleter = &DlpackDeleter;
+  ort_dlmanaged_tensor->tensor.dl_tensor.data = const_cast<void*>(tensor.DataRaw());
+  ort_dlmanaged_tensor->tensor.dl_tensor.ctx = GetDlpackContext(ort_value, tensor.Location().device.Id());
+  ort_dlmanaged_tensor->tensor.dl_tensor.ndim = static_cast<int>(tensor.Shape().NumDimensions());
+  ort_dlmanaged_tensor->tensor.dl_tensor.dtype = GetDlpackDataType(ort_value);
+  ort_dlmanaged_tensor->tensor.dl_tensor.shape =
+      tensor.Shape().NumDimensions() > 0 ? const_cast<int64_t*>(&tensor.Shape()[0]) : nullptr;
+  ort_dlmanaged_tensor->tensor.dl_tensor.strides = nullptr;
+  ort_dlmanaged_tensor->tensor.dl_tensor.byte_offset = 0;
+  return &(ort_dlmanaged_tensor->tensor);
+}
+
+OrtValue DlpackToOrtValue(const DLManagedTensor* dlpack) {
+  // ORT only supports contiguous tensor for now.
+  ORT_ENFORCE(IsContiguousTensor(dlpack->dl_tensor), "ORT only supports contiguous tensor for now.");
+  OrtDevice device = GetOrtDevice(dlpack->dl_tensor.ctx);
+  MLDataType data_type = GetOrtValueDataType(dlpack->dl_tensor.dtype);
+  std::function<void(void*)> deleter = [dlpack](void*) { dlpack->deleter(const_cast<DLManagedTensor*>(dlpack)); };
```

Comment:
>const_cast [](start = 73, length = 10)

similar question about why we need const_cast. does dlpack need to be a const DLManagedTensor*?

Reply:
dlpack should not be const here. updated.